src/backends/file-system/API.js

```diff
@@ -0,0 +1,166 @@
+      if (contentType && contentType.match(/json/)) {
+        return this.parseJsonResponse(response);
+      }
+      return response.text();
+    })
+    .catch((error) => {
+      throw new APIError(error.message, responseStatus, 'fs');
+    });
+  }
+
+  readFile(path) {
+    const cache = Promise.resolve(null);
+    return cache.then((cached) => {
+      if (cached) { return cached; }
+
+      return this.request(`/file/${ path }`, {
+        headers: { Accept: "application/octet-stream" },
+        params: { },
+        cache: "no-store",
+      }).then((result) => {
+        return result;
+      });
+    });
+  }
+
+  listFiles(path) {
+    return this.request(`/files/${ path }`, {
+      params: { },
+    })
+    .then((files) => {
+      if (!Array.isArray(files)) {
+        throw new Error(`Cannot list files, path ${ path } is not a directory but a ${ files.type }`);
+      }
+      return files;
+    })
+    .then(files => files.filter(file => file.type === "file"));
+  }
+
+  composeFileTree(files) {
+    let filename;
+    let part;
+    let parts;
+    let subtree;
+    const fileTree = {};
+
+    files.forEach((file) => {
+      if (file.uploaded) { return; }
+      parts = file.path.split("/").filter(part => part);
+      filename = parts.pop();
+      subtree = fileTree;
+      while (part = parts.shift()) {
+        subtree[part] = subtree[part] || {};
+        subtree = subtree[part];
+      }
+      subtree[filename] = file;
+      file.file = true;
+    });
+
+    return fileTree;
+  }
+
+  toBase64(str) {
+    return Promise.resolve(
+      Base64.encode(str)
+    );
+  }
+
+  uploadBlob(item, newFile = false) {
+    const content = item instanceof AssetProxy ? item.toBase64() : this.toBase64(item.raw);
+    const method = newFile ? "POST" : "PUT"; // Always update or create new. PUT is Update existing only
+
+    const pathID = item.path.substring(0, 1) === '/' ? item.path.substring(1, item.path.length) : item.path.toString();
+
+    return content.then(contentBase64 => this.request(`/file/${ pathID }`, {
+      method: method,
+      body: JSON.stringify({
+        content: contentBase64,
+        encoding: "base64",
+      }),
+    }).then((response) => {
+      item.uploaded = true;
+      return item;
+    }));
+  }
+
+  persistFiles(entry, mediaFiles, options) {
+    const uploadPromises = [];
+    const files = mediaFiles.concat(entry);
+
+    files.forEach((file) => {
+      if (file.uploaded) { return; }
+      uploadPromises.push(this.uploadBlob(file, (options.newEntry && !(file instanceof AssetProxy))));
+    });
+
+    const fileTree = this.composeFileTree(files);
+
+    return Promise.all(uploadPromises).then(() => {
+      if (!options.mode || (options.mode && options.mode === SIMPLE)) {
+        return fileTree;
+      }
+    });
+  }
+
+  deleteFile(path, message, options={}) {
+    const fileURL = `/file/${ path }`;
+    return this.request(fileURL, {
+      method: "DELETE",
+      params: { },
```

Comment:
These `params: { },` should be necessary.

Reply:
right now there is nothing being done with them to the file-system, but maybe later in the case we want to do logging etc in next phase.